### PR TITLE
[FLINK-22376][runtime] RecoveredChannelStateHandler recycles the buffer if it was created inside and doesn't recycle if it was passed as parameter.

### PR DESCRIPTION
**Pull request by akalash**
## What is the purpose of the change

*PR organizes the order of recycling of buffer inside of RecoveredChannelStateHandler *

## Brief change log
  - *BufferBuilder is able to recycle itself if there is no consumer was created*
  - *InputChannelRecoveredStateHandler doesn't recycle buffer inside of recover*
  - *ResultSubpartitionRecoveredStateHandler create buffer consumer inside recover rather than receiving it as parameter*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added tests for BufferBuilder, InputChannelRecoveredStateHandler, ResultSubpartitionRecoveredStateHandler*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/BufferBuilder.java [hunk 1]
```diff
@@ -153,7 +153,12 @@ public BufferRecycler getRecycler() {
     }
 
     public void recycle() {
-        recycler.recycle(memorySegment);
+        // If at least one consumer was created then they responsible for the memory recycling
+        // because BufferBuilder doesn't contain a references counter so it will be impossible to
+        // correctly recycle memory here.
+        if (!bufferConsumerCreated) {
+            recycler.recycle(memorySegment);
+        }
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/BufferBuilder.java:161]: Frankly, this seems (still?) like a partial solution/hack, because what's the contract when this method should be called? Before this `recycle()` method was introduced, it was at least clear, that `BufferBuilder` is never recycling the segment, and this is always done via closing `BufferConsumer`s. Now it seems like we are going deeper into the murky waters that "sometimes" `recycle()` should be called?

I think I would like the "hack" of relaying on the `bufferConsumerCreated` flag to avoid retaining the buffer in the `BufferBuilder`, but I think it's still confusing:
1. It would be probably better to make `BufferBuilder` implement `Closeable` and rename `recycle()` to `close()` - this would probably cause quite a bit of changes, especially in tests.
2. It still doesn't solve the problem of writing to already released `memorySegment`, like what if the `BufferConsumer` was created and has already been closed, while someone is still writing data to the `BufferBuilder`? There was a bug like that, that I fixed on a different layer, but maybe we should fix this after all here as well? I mean especially measuring on the benchmarking machine if there is any overhead of retaining and recycling `NetworkBuffer` one extra time?

**Review by pnowojski — COMMENTED** [review]
Thanks for the changes @akalash , I've left two questions.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/BufferBuilder.java (hunk 1 above)
Comment (akalash) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/BufferBuilder.java:161]: Yes, it is still definitely a hack. In my opinion, the right solution is to avoid direct writing into `MemorySegment`  from `BufferBuilder`. It means we just should change the implementation of `BufferBuilder` in such a way that using `Buffer` instead of `MemorySegment`. As I understand now you don't have any objections about such a solution if the benchmark doesn't show any degradation?

In any case, some answers to the questions:

- The contract is simple - `BufferBuilder#recycle()` should be called always when `BufferBuilder` is not needed anymore. You don't need to think `BufferConsumer` was created or not.
- In general, renaming `recycle()` to `close()` makes sense to me since `BufferBuilder` doesn't have `retain` method and ideally, should be closed after usage.(we can think about it when we will agree on a final solution)
- There are a couple of problems still not resolved - writing to already released `memorySegment` or creating 'BufferConsumer' from already closed 'BufferBuilder'. They both can be resolved by solution which we already discussed(using `Buffer` instead of `memorySegment` inside of `BufferBuilder`)

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RecoveredInputChannel.java [hunk 2]
```diff
@@ -120,38 +120,36 @@ public void checkpointStopped(long checkpointId) {
...
-                buffer.recycleBuffer();
+        final boolean wasEmpty;
+        synchronized (receivedBuffers) {
+            // Similar to notifyBufferAvailable(), make sure that we never add a buffer
+            // after releaseAllResources() released all buffers from receivedBuffers.
+            if (isReleased) {
+                wasEmpty = false;
+            } else {
+                wasEmpty = receivedBuffers.isEmpty();
+                receivedBuffers.add(buffer.retainBuffer());
```
[the remark this answers is not in the document]
Comment (akalash) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RecoveredInputChannel.java:137]: Just in case, a little spoiler - these changes are not a target of the current task so I can easily revert these changes.

Of course, it is not ok, that there are inconsistent between different parts of the code. So I will change everything in the same manner when we agree with the solution(or will do nothing if we decide to leave everything as is). 

According to the original question. As I see right now there is no clear pattern of usage retain/release. It is why I want to find something which will be understandable for everybody. In my opinion, the classic pattern of reference/resource counter looks like the following: 

- if you request resources no matter what happens next you should release this resource. 
- If you share a resource you should increase the counter(ideally, you are able to share only via a specific structure that does all work for you. ex. BufferList instead of List<Buffer> which retain buffer automatically)
- If you passive consumer which receives resources from outside you should do nothing with the management of resources.

```
void wrong() {
  Buffer buf = getOrCreate();//request resource without release
  ...
}
void right() {
  Buffer buf = getOrCreate();//request resource
  try {
    ...
  } finally {
   buf.release();
 }
}
```

```
void wrong(Buffer buf) {
  buf.relase();//how I can be sure that it is enough of counter.
  
  buf.retain();//retain without release, can be surprising for the caller of this method.
}
void right(Buffer buf) {
  buf.retain();//useless but possible
  try {
    ...
  } finally {
   buf.release();
 }
}
```

```
void wrong(Buffer buf) {
  Buffer buf = getOrCreate();//request resource
  list.add(buf);//too dangerous. can lead to leak after refactoring. but possible in extreme cases for good performance.
}
void right(Buffer buf) {
  Buffer buf = getOrCreate();//request resource
  try {
    list.add(buf.retain()); //it is better to have BufferList with automatic 'retain' but this is also possible.
  } finally {
   buf.release();
 }
}
```
```
void right(Buffer buf) {
  list.add(buf.retain()); //it is better to have BufferList with automatic 'retain' but this is also possible.
  
  buf.retain(); //again, it is better to have something more specific for sharing resources with other threads but this also works.
  executor.submit(()-> /* some work with buffer*/);
}
```

So let's discuss what do you think about it. And maybe you can provide your rules of using retain/release pattern which I didn't realize from the code.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/BufferBuilder.java (hunk 1 above)
Comment (akalash) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/BufferBuilder.java:161]: I have investigated the code a little more and now I understand that it is not so easy to rewrite `BufferBuilder`(at least I see pretty strange usage of `BufferBuilder` inside of `PartitionSortedBuffer` and `SortMergeResultPartition`). So I can propose for this PR to remove the first commit because it is too controversial and merge the second commit as is - it is not the solution for all problems but it is resolving one of them. And I also suggest creating a ticket where we can think about the correct usage of `MemorySegment` and remove it from the places where it doesn't make sense in order to keep the contract simple and consistent.
Also, if we think that this ticket is not so urgent, we can discuss `MemorySegment` usage here rather than in a separate ticket.
 WDYT? @pnowojski

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/recovery/DemultiplexingRecordDeserializerTest.java [hunk 3]
```diff
@@ -190,10 +196,12 @@ public void testWatermarks() throws IOException {
                         deserializer.getVirtualChannelSelectors().iterator();
                 iterator.hasNext(); ) {
             SubtaskConnectionDescriptor selector = iterator.next();
-            final BufferBuilder bufferBuilder = createBufferBuilder(128);
+            MemorySegment memorySegment = allocateUnpooledSegment(128);
+            final BufferBuilder bufferBuilder = createBufferBuilder(memorySegment);
             final long ts =
                     42L + selector.getInputSubtaskIndex() + selector.getOutputSubtaskIndex();
             Buffer buffer = write(bufferBuilder, new Watermark(ts));
+            bufferBuilder.recycle();
```
Comment (pnowojski) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/recovery/DemultiplexingRecordDeserializerTest.java]: semi nit:
In quite a bit places, you could have used
```
try (BufferBuilder bufferBuilder = ...) {
}
```
For this, it would be easier to first refactor `BufferBuilder` to implement `AutoCloseable` as a first commit, not the last one. I'm not entirely sure if this is worth the effort.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BufferWritingResultPartition.java [hunk 4]
```diff
@@ -348,6 +348,7 @@ private void finishUnicastBufferBuilder(int targetSubpartition) {
         if (bufferBuilder != null) {
             numBytesOut.inc(bufferBuilder.finish());
             numBuffersOut.inc();
+            bufferBuilder.recycle();
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BufferWritingResultPartition.java]: do we need both `finish()` and `recycle()`? Wouldn't single `close()` be enough?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandler.java [hunk 5]
```diff
@@ -85,19 +84,15 @@
...
+                        EventSerializer.toBuffer(
+                                new SubtaskConnectionDescriptor(
+                                        oldSubtaskIndex, channelInfo.getInputChannelIdx()),
+                                false));
+                channel.onRecoveredStateBuffer(buffer.retainBuffer());
             }
-        } finally {
-            buffer.recycleBuffer();
         }
     }
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandler.java:97]: 1. I don't like that we are mixing different styles of contracts here. `RecoveredChannelStateHandler#recover` receives a `Buffer` that it doesn't own - caller is responsible for the clean up. While a couple lines below its copying buffer and passing the ownership of the copy to the `channel` - callee is responsible for the clean up. Why not stick with something that's consistent, and just have a contract that we are passing the ownership of the buffer to the callee? 
2. It's not documented anywhere.

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        buffer.buffer.recycle();
+
+        // then: Buffer should be recycled the same times as it was retained.
+        stateHandlerTester.assertState();
+    }
+
+    @Test
+    public void testRecycleBufferAfterRecoverWasCalled() throws Exception {
+        // given: Initialized state handler.
+        stateHandlerTester.init();
```
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandlerTest.java:86]: `@Before`?

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+        Info channelInfo = stateHandlerTester.getInfo();
+
+        // when: Request the buffer.
+        RecoveredChannelStateHandler.BufferWithContext<Context> buffer =
+                stateHandlerTester.getBuffer(channelInfo);
+
+        // and: Recycle buffer outside.
+        buffer.buffer.recycle();
```
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandlerTest.java:77]: `buffer.buffer.`?

`%s/buffer/bufferWithContext/g`?

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandlerTest.java [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+import java.io.IOException;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.junit.Assert.assertEquals;
+
+/** Test of different implementation of {@link RecoveredChannelStateHandler}. */
+@RunWith(Parameterized.class)
+public class RecoveredChannelStateHandlerTest<Info, Context> {
```
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandlerTest.java:49]: I'm having quite hard time to follow this test.

I have a feeling that parametrising this test did more harm than good?

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandlerTest.java [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+        // and: Recycle buffer outside.
+        buffer.buffer.recycle();
+
+        // then: Buffer should be recycled the same times as it was retained.
+        stateHandlerTester.assertState();
+    }
+
+    static class ResultSubpartitionRecoveredChannelStateHandlerTester
+            extends RecoveredChannelStateHandlerTester<ResultSubpartitionInfo, BufferBuilder> {
+        AtomicInteger recycledMemorySegments;
```
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandlerTest.java:106]: Why `AtomicInteger`?

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandlerTest.java [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+            partition =
+                    new ResultPartitionBuilder()
+                            .setNetworkBufferPool(
+                                    new NetworkBufferPool(3, 32 * 1024) {
+                                        @Override
+                                        public void recycle(MemorySegment segment) {
+                                            recycledMemorySegments.incrementAndGet();
+                                            super.recycle(segment);
+                                        }
+                                    })
```
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandlerTest.java:127]: Why not use plain simple `NetworkBufferPool` and just assert at the end the number of available segments instead of counting how many times `recycle` was called? Frankly, the test shouldn't care how many times `recycle` was called. Just that the buffers were returned to the pool.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/util/TestPooledBufferProvider.java [hunk 11]
```diff
@@ -36,9 +36,10 @@
 
 import static org.apache.flink.util.Preconditions.checkArgument;
 
+/** */
 public class TestPooledBufferProvider implements BufferProvider {
 
-    private final BlockingQueue<Buffer> buffers = new LinkedBlockingDeque<>();
+    private final BlockingQueue<MemorySegment> buffers = new LinkedBlockingDeque<>();
```
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/util/TestPooledBufferProvider.java:42]: segments?

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/util/TestPooledBufferProvider.java [hunk 12]
```diff
@@ -132,26 +146,24 @@ public int getNumberOfCreatedBuffers() {
 
         private final Object listenerRegistrationLock = new Object();
 
-        private final Queue<Buffer> buffers;
+        private final Queue<MemorySegment> buffers;
```
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/util/TestPooledBufferProvider.java:149]: segments?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java [hunk 13]
```diff
@@ -172,8 +172,7 @@ public void setup() throws IOException {
         synchronized (lock) {
             try {
                 for (int i = 0; i < numWriteBuffers; ++i) {
-                    MemorySegment segment =
-                            bufferPool.requestBufferBuilderBlocking().getMemorySegment();
+                    MemorySegment segment = bufferPool.requestMemorySegmentBlocking();
                     writeBuffers.add(segment);
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java:176]: writeSegments?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PartitionSortedBuffer.java [hunk 14]
```diff
@@ -270,14 +269,13 @@ private MemorySegment requestBufferFromPool() throws IOException {
         try {
             // blocking request buffers if there is still guaranteed memory
             if (buffers.size() < numGuaranteedBuffers) {
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PartitionSortedBuffer.java:271]: segments?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateSerializer.java [hunk 15]
```diff
@@ -103,7 +103,7 @@ public boolean isWritable() {
 
             @Override
             public void recycle() {
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateSerializer.java:105]: `close()`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BufferWritingResultPartition.java [hunk 16]
```diff
@@ -363,7 +363,7 @@ private void finishBroadcastBufferBuilder() {
         if (broadcastBufferBuilder != null) {
             numBytesOut.inc(broadcastBufferBuilder.finish() * numSubpartitions);
             numBuffersOut.inc(numSubpartitions);
-            broadcastBufferBuilder.recycle();
+            broadcastBufferBuilder.close();
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BufferWritingResultPartition.java:366]: You are modifying a code that you have introduced in the same PR (increasing PR size). It would be better to have reversed order of the commits.

**Review by pnowojski — COMMENTED** [review]
Thanks for the changes @akalash . I have left some comments. Have you benchmarked this code? Could I see a comparison against the master?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/recovery/DemultiplexingRecordDeserializerTest.java (hunk 3 above)
Comment (akalash) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/recovery/DemultiplexingRecordDeserializerTest.java]: Unfortunately, it is not possible to implement `AutoCloseable` as the first commit because `BufferBuilder#recycle` works correctly only after my changes. I mean this code:
```
try(BufferBuilder bb = requestBufferBuilder) {
   return bb.createBufferConsumer();
}
```
is wrong because we return an already recycled buffer. But after my changes, it works fine. 
Of course, I can move this commit to the first place for easing reviewing but this commit will be incorrect without the next ones.

P.S. I wrapped a couple more places with try-with-resources.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BufferWritingResultPartition.java (hunk 4 above)
Comment (akalash) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BufferWritingResultPartition.java]: Yes, unfortunately, we need both because in case of calling `close()` on exception we want to recycle buffer but don't want to finish it. In fact, I see only one place where we definitely should distinguish these methods so maybe it will be possible to reduce it to one method but I am still not sure.

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandlerTest.java (hunk 8 above)
Comment (akalash) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandlerTest.java:41]: I rewrote the tests

**Comment by akalash** [comment]
@pnowojski You can find benchmarks result here http://codespeed.dak8s.net:8080/job/flink-benchmark-request/291/ or http://codespeed.dak8s.net:8000/comparison/ (until nobody rewrite it)

**Comment by akalash** [comment]
@flinkbot run azure

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BufferWritingResultPartition.java (hunk 4 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BufferWritingResultPartition.java]: Can you point to what is this place where it matters? From the top of my head it seems like it shouldn't be an issue.

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandlerTest.java [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+
+import java.util.HashSet;
+
+import static org.junit.Assert.assertEquals;
+
+/** Test of different implementation of {@link InputChannelRecoveredStateHandler}. */
+public class RecoveredChannelStateHandlerTest<Info, Context> {
+
+    @Test
+    public void testRecycleBufferForInputChannelBeforeRecoverWasCalled() throws Exception {
```
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandlerTest.java:46]: optional comment:

If you want to avoid some of the code duplication, maybe we can split this test into two test classes. `RecoveredChannelStateHandlerTest` -> `InputRecoveredChannelStateHandlerTest` and `OutputRecoveredChannelStateHandlerTest`. 

Both input and output variants could do initialisation in the `@Before` methods for example:
```
private NetworkBufferPool networkBufferPool;
private SingleInputGate inputGate;
private InputChannelRecoveredStateHandler icsHander;

@Before
public void setUp() {
        networkBufferPool = new NetworkBufferPool(preAllocatedSegments, 1024);

        // and: Configured input gate with recovered channel.
        inputGate =
                new SingleInputGateBuilder()
                        .setChannelFactory(InputChannelBuilder::buildLocalRecoveredChannel)
                        .setSegmentProvider(networkBufferPool)
                        .build();

        icsHander = buildInputChannelStateHandler(inputGate);  
}

@Test
public void testRecycleBufferForInputChannelBeforeRecoverWasCalled() { ...}


@Test
public void testRecycleBufferForInputChannelAfterRecoverWasCalled() {...}


```
?

**Review by pnowojski — COMMENTED** [review]
Thanks for the update. I've left one question and one more comment. But I think the code looks good to me :)

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandler.java [hunk 18]
```diff
@@ -83,8 +91,12 @@
...
 
     @Override
-    public void recover(InputChannelInfo channelInfo, int oldSubtaskIndex, Buffer buffer)
+    public void recover(
```
[the remark this answers is not in the document]
Comment (akalash) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandler.java:94]: I actually have almost the same comment on the interface method I just decided to not duplicate it on each implementation. But it is actually good idea to add some comment on the onBuffer

[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandler.java:94]: Ops, I missed that java doc :) Yes, adding it to the `onBuffer` would be nice.

**Comment by akalash** [comment]
I pushed the last changes(test refactoring, AutoClosable) into existing commits and I added one more minor commit for RemoteInputChannel javadoc

**Comment by pnowojski** [comment]
LGTM % green azure, and one remaining question.

Yesterday there was a [test failure](https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=18521&view=logs&j=0da23115-68bb-5dcd-192c-bd4c8adebde1&t=05b74a19-4ee4-5036-c46f-ada307df6cf0) that looked very relevant to your changes. Have you noticed/fixed it? If not, we shouldn't ignore it, even if it didn't occur in later builds, as this can be some intermittent bug/race condition.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BufferWritingResultPartition.java (hunk 4 above)
Comment (akalash) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BufferWritingResultPartition.java]: at least here org.apache.flink.runtime.checkpoint.channel.ChannelStateChunkReader#readChunk. 
And for example here org.apache.flink.runtime.checkpoint.channel.ResultSubpartitionRecoveredStateHandler#recover, right now we create BufferConsumer and only then finishing the BufferBuilder and it is possible to safely replace the finish by the close but I think more correctly in this method to finish BufferBuilder first, then create BufferConsumer and only then close BufferBuilder.

But in general, I believe it is possible to reduce these two methods to one (it is not a big problem even for readChunk because BufferConsumer is not created yet). But in this case, we will leave only one possible invariant - commit all changes only when BufferConsumer is created already. So filling the buffer then committing changes and only then creating BufferConsumer will be impossible.
So for example this code won't be work:
```
try(BufferBuilder bb = requestBufferBuilder()) {
   bb.append(..);
   read(bb.createBufferConsumer());
}
```
it will need to be rewritten like this:
```
BufferBuilder bb = requestBufferBuilder();
bb.append(..);
BufferConsumer bc = bb.createBufferConsumer();
bb.close();
read(bc);
```
In conclusion, I think that logically it is two different methods(different responsibilities), but for simplification, it is possible to have only one method if you think it makes sense.

**Comment by akalash** [comment]
@pnowojski, I didn't ignore it, I fixed it(it was problem with test which I broke earlier)
